File: week3/code/get_TreeHeight.R
#!/usr/bin/Rscript 
#get_TreeHeight.R

rm(list=ls())

# to do:
#includes the input file name in the output file name as InputFileName_treeheights.csv. 
# remove the prefix and the suffix from the file to get only the name 
args = commandArgs(trailingOnly=TRUE) # function to scan the arguments supplied to R 

if (length(args) == 0) {
  print("No input file was given: using trees.csv as a default file")
  args[1] <- "../data/trees.csv"
} else if (length(args) >1){
  print("too many inputs were provided: using trees.csv as a default file")
  args[1] <- "../data/trees.csv" 
} else if (length(args) == 1 && file.exists(args[1])){
  print(paste("The input file is:", args[1]))
} else{
   print("we can not find the input file. Using trees.csv")
   args[1] <- "../data/trees.csv"
}


Trees <- read.csv(args[1], header = TRUE) #reading the csv of args[1]

#degrees <- Trees$Angle.degrees #puts angle in degrees
#distance <- Trees$Distance.m # puts distances in distance

TreeHeight <- function(degrees, distance){
 radians <- degrees * pi / 180 #this i to convert the degrees to radian
  height <- distance * tan(radians)
  return (height)
}

Trees$Tree.Height.m <- TreeHeight(Trees$Angle.degrees, Trees$Distance.m) #runs the function and rint output in TreeHeight
#Trees # to display the table in bash 

#write.csv(Trees, "../results/trees.csv", row.names  =F)#write in new file

#outputfilename <- paste0(basename(tools::file_path_sans_ext(args[1])),"_treeheights.csv") 

# removing the extansion with the tools::file_path_
# paste0 is like paste() but it has sep = "" as default argument , so any string gets merged together

#write.csv(Trees, paste0(tools::file_path_sans_ext(args[1]),"_treeheights.csv"), "../results")
write.csv(Trees, paste0("../results/",basename(tools::file_path_sans_ext(args[1])),"_treeheights.csv"),row.names = FALSE)

#output_folder <- paste0("../results/",outputfilename)
#write.csv(Trees, output_folder)
print("Script complete!")

File: week3/code/get_TreeHeight.py
#!/usr/bin/env python3

"""
Author: Group3
Script: get_TreeHeight.py
Des: Calculate the treeheight according to the degrees and distances
Usage: python3 get_TreeHeight.py file
Dep: pandas,math,sys,os
Date: Oct, 2021
"""

__appname__ = '[get_TreeHeight.py]'
__author__ = 'Group3'
__version__ = '0.0.1'

import pandas as pd #data ana
import math #use math formula
import sys #interact with the system
import os #deal with the directory and file

def TreeHeight(degrees, distance):

    """
    Args:
        degrees: 
        distance: 
    Returns:
        Height
    Des:
        calculate the Height by degrees and distance provided
    
    """
    radians = degrees * math.pi / 180 #
    #radians = math.radians(degrees)
    tans = radians.apply(lambda x: math.tan(x))
    height = distance * tans
    return height

def main(argv):

    """ 
    Main process running the program.
    """ 
    if len(sys.argv) == 2:
        try:
            MyData = pd.read_csv(sys.argv[1]) 
            infilename = os.path.basename(sys.argv[1]).split(sep=".")[0]
        except (FileNotFoundError):
            print ("Your files provided here are not accessible")
            sys.exit(0)
    else:
        print ("We will use the default path here\n ")
        path1 = "../data/trees.csv"
        infilename = os.path.basename(path1).split(sep=".")[0]
        MyData = pd.read_csv(path1)

    outputfile = "../results/"+infilename+"_treeheights.csv"

    MyData["Tree.Height.m"] = TreeHeight(MyData["Angle.degrees"],MyData["Distance.m"])
    MyData.to_csv(outputfile,index=False)    
    return 0

if (__name__ == "__main__"):
    print ("We are now running:",sys.argv[0],"\n")
    status = main(sys.argv)
    sys.exit(status)



